### PR TITLE
Fx barrier option: trade model and Black pricer

### PR DESCRIPTION
**Pull request by yukiiwashita**

### CODE REVIEW

File: modules/product/src/main/java/com/opengamma/strata/product/fx/Barrier.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * Please see distribution for license.
+ */
+package com.opengamma.strata.product.fx;
+
+import java.time.LocalDate;
+
+/**
+ * Definition of barrier event of option instruments. 
+ * <p>
+ * The barrier type, knock type, barrier level and other relevant information are specified in this class.
```
Comment (marc-opengamma) [inline modules/product/src/main/java/com/opengamma/strata/product/fx/Barrier.java]: The barrier level can be date dependent. For forward starting barrier, the barrier level can be set to very high or low level in the initial period.

File: modules/product/src/main/java/com/opengamma/strata/product/fx/FxSingleBarrierOption.java [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+   * <p>
+   * The barrier level stored in this field must be represented based on the direction of the currency pair in the 
+   * underlying FX transaction. 
+   * <p>
+   * For example, if the underlying option is an option on EUR/GBP, the barrier should be a certain level of EUR/GBP rate.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final Barrier barrier;
+  /**
+   * The amount paid back to the option holder in case the option becomes inactive.
```
Comment (marc-opengamma) [inline modules/product/src/main/java/com/opengamma/strata/product/fx/FxSingleBarrierOption.java]: For a 'out' option, the amount is paid when the barrier is reached; for a 'in' option, the amount is paid at expiry if the barrier is not reached.

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/BlackFxSingleBarrierOptionProductPricer.java [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+   * the underlying curves.
+   * <p>
+   * The volatility is fixed in this sensitivity computation.
+   * 
+   * @param option  the option product
+   * @param ratesProvider  the rates provider
+   * @param volatilityProvider  the Black volatility provider
+   * @return the present value curve sensitivity of the product
+   */
+  public PointSensitivityBuilder presentValueSensitivity(
```
Comment (marc-opengamma) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/BlackFxSingleBarrierOptionProductPricer.java]: We should probably rename the method to "presentValueSensitivityStickyStrike" (same in "BlackFxVanillaOptionProductPricer") to be coherent with "VolatilitySwaptionPhysicalProductPricer" and to allow the introduction of the 'SensitivityStickyMoneyness' and 'SensitivityStickyDelta' later. 
Add also in the Java doc the fact that this is 'StickyStrike' sensitivity.

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/BlackFxSingleBarrierOptionProductPricer.java [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+    CurrencyPair currencyPair = underlyingFx.getCurrencyPair();
+    DiscountFactors baseDiscountFactors = ratesProvider.discountFactors(ccyBase);
+    DiscountFactors counterDiscountFactors = ratesProvider.discountFactors(ccyCounter);
+
+    double rateBase = baseDiscountFactors.zeroRate(underlyingFx.getPaymentDate());
+    double rateCounter = counterDiscountFactors.zeroRate(underlyingFx.getPaymentDate());
+    double costOfCarry = rateCounter - rateBase;
+    double dfBase = baseDiscountFactors.discountFactor(underlyingFx.getPaymentDate());
+    double dfCounter = counterDiscountFactors.discountFactor(underlyingFx.getPaymentDate());
+    double spot = ratesProvider.fxRate(currencyPair);
```
Comment (marc-opengamma) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/BlackFxSingleBarrierOptionProductPricer.java]: Probably better to call if "todayFx" as "spot" would refer to T+2 value.

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/BlackFxSingleBarrierOptionProductPricer.java [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+    return CurrencyAmount.of(underlyingOption.getCounterCurrency(), signedNotional(underlyingOption) * price);
+  }
+
+  /**
+   * Calculates the price of the FX barrier option product.
+   * <p>
+   * The price of the product is the value on the valuation date for one unit of the base currency 
+   * and is expressed in the counter currency. The price does not take into account the long/short flag. 
+   * See {@link #presentValue} for scaling and currency.
+   *
```
Comment (marc-opengamma) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/BlackFxSingleBarrierOptionProductPricer.java]: Add explanation on how the volatility is selected. As there is a strike and a barrier, it is not obvious which number is used.

File: modules/pricer/src/test/java/com/opengamma/strata/pricer/fx/BlackFxSingleBarrierOptionProductPricerTest.java [hunk 6]
```diff
@@ -0,0 +1,678 @@
...
+      CurrencyAmount pv = PRICER.presentValue(option, RATE_PROVIDER, VOL_PROVIDER);
+      MultiCurrencyAmount computed = PRICER.currencyExposure(option, RATE_PROVIDER, VOL_PROVIDER);
+      FxMatrix fxMatrix = FxMatrix.builder().addRate(EUR, USD, SPOT + FD_EPS).build();
+      ImmutableRatesProvider provBumped = RATE_PROVIDER.toBuilder().fxRateProvider(fxMatrix).build();
+      CurrencyAmount pvBumped = PRICER.presentValue(option, provBumped, VOL_PROVIDER);
+      double ceCounterFD = pvBumped.getAmount() - pv.getAmount();
+      double ceBaseFD = pvBumped.getAmount() * (SPOT + FD_EPS) - pv.getAmount() * SPOT;
+      assertEquals(computed.getAmount(EUR).getAmount() * FD_EPS, ceCounterFD, NOTIONAL * FD_EPS);
+      assertEquals(-(computed.getAmount(USD).getAmount() - pv.getAmount()) * FD_EPS, ceBaseFD, NOTIONAL * FD_EPS);
+    }
```
Comment (marc-opengamma) [inline modules/pricer/src/test/java/com/opengamma/strata/pricer/fx/BlackFxSingleBarrierOptionProductPricerTest.java]: The last 3 lines of the test are incorrect. Replace by:

      double ceBaseFD = pvBumped.getAmount() / (SPOT + FD_EPS) - pv.getAmount() / SPOT;
      assertEquals(computed.getAmount(EUR).getAmount() * FD_EPS, ceCounterFD, NOTIONAL * TOL);
      assertEquals(computed.getAmount(USD).getAmount() * (1.0d/ (SPOT + FD_EPS) - 1.0d/ SPOT), ceBaseFD, NOTIONAL * TOL);

File: modules/pricer/src/test/java/com/opengamma/strata/pricer/fx/BlackFxSingleBarrierOptionProductPricerTest.java [hunk 7]
```diff
@@ -0,0 +1,678 @@
...
+  }
+
+  public void test_currencyExposure_afterExpiry() {
+    for (ResolvedFxSingleBarrierOption option : OPTION_ALL) {
+      MultiCurrencyAmount computed = PRICER.currencyExposure(option, RATE_PROVIDER_AFTER, VOL_PROVIDER_AFTER);
+      assertEquals(computed, MultiCurrencyAmount.empty());
+    }
+  }
+
```
Comment (marc-opengamma) [inline modules/pricer/src/test/java/com/opengamma/strata/pricer/fx/BlackFxSingleBarrierOptionProductPricerTest.java]: Add a test on currency exposure comparing vanilla CE with barrier with a very large or very low barrier. This is done for the price indirectly in the Barrier formula repository test, but not for CE.

Comment (yukiiwashita) [inline modules/pricer/src/test/java/com/opengamma/strata/pricer/fx/BlackFxSingleBarrierOptionProductPricerTest.java]: Added farBarrierOutTest and farBarrierInTest.

**Pull request merged**